### PR TITLE
Improvements to FormatMessage

### PR DESCRIPTION
**Pull request by stephentoub**
This PR does two things.

First, it unifies our copies of Interop.FormatMessage.cs between the shared copy from CoreLib and the corefx copy; we now just use the one shared with CoreLib, but replaced with the one from corefx that had a few improvements from https://github.com/dotnet/corefx/commit/98227f1a18c1859c5614a0f045a09660f34e6cc1.

Second, it fixes a long-standing annoyance where error messages from Win32 have their ending period removed.  I don't know why this was done, as it's been there effectively forever, but there does not seem to be a good reason for it.  Example:
```C#
using System;
using System.ComponentModel;

class Test
{
    public static void Main()
    {
        for (int i = 1; i < 20; i++) Console.WriteLine(new Win32Exception(i).Message);
    }
}
```
Before:
```
Incorrect function
The system cannot find the file specified
The system cannot find the path specified
The system cannot open the file
Access is denied
The handle is invalid
The storage control blocks were destroyed
Not enough memory resources are available to process this command
The storage control block address is invalid
The environment is incorrect
An attempt was made to load a program with an incorrect format
The access code is invalid
The data is invalid
Not enough memory resources are available to complete this operation
The system cannot find the drive specified
The directory cannot be removed
The system cannot move the file to a different disk drive
There are no more files
The media is write protected
```
After:
```
Incorrect function.
The system cannot find the file specified.
The system cannot find the path specified.
The system cannot open the file.
Access is denied.
The handle is invalid.
The storage control blocks were destroyed.
Not enough memory resources are available to process this command.
The storage control block address is invalid.
The environment is incorrect.
An attempt was made to load a program with an incorrect format.
The access code is invalid.
The data is invalid.
Not enough memory resources are available to complete this operation.
The system cannot find the drive specified.
The directory cannot be removed.
The system cannot move the file to a different disk drive.
There are no more files.
The media is write protected.
```

cc: @jkotas, @danmosemsft

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet/dnceng, @DrewScoggins, the OSX leg is failing across multiple PRs with:
https://ci3.dot.net/job/dotnet_corefx/job/master/job/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/12177/
```
requests.exceptions.HTTPError : 500 Server error : Internal Server Error [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/tests.builds]
  downloaded dumpling.py
  total elapsed time 0:00:00.184697
  {
      "dbgpath": "~/.dumpling/dbg/bin/lldb",
      "installpath": "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/tempHome/.dumpling",
      "url": "https://dumpling.int-dot.net/"
  }
  configuration saved to /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/tempHome/.dumpling/dumpling.config.json
  total elapsed time 0:00:00.004976
/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/Dumpling.targets(30,5): error MSB3073: The command "python /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/DumplingHelper.py install_dumpling" exited with code -1. [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/tests.builds]
```

File: src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs [hunk 1]
```diff
@@ -38,50 +39,51 @@ internal static string GetMessage(int errorCode)
...
                     return errorMsg;
                 }
-                else
-                {
-                    // increase the capacity of the StringBuilder.
-                    sb.Capacity *= BufferSizeIncreaseFactor;
-                }
+
+                // Increase the capacity of the buffer.
+                buffer = new char[buffer.Length * BufferSizeIncreaseFactor];
```
Comment (danmosemsft) [inline src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: Array pool?

File: src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs [hunk 2]
```diff
@@ -14,17 +13,19 @@ internal partial class Kernel32
         private const int FORMAT_MESSAGE_FROM_HMODULE = 0x00000800;
         private const int FORMAT_MESSAGE_FROM_SYSTEM = 0x00001000;
         private const int FORMAT_MESSAGE_ARGUMENT_ARRAY = 0x00002000;
-
-
+        
         private const int ERROR_INSUFFICIENT_BUFFER = 0x7A;
+        private const int InitialBufferSize = 256; // small enough to be on stack, and large enough for most error messages
```
Comment (danmosemsft) [inline src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: #if DEBUG constants to exercise expansion path?

File: src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs [hunk 3]
```diff
@@ -38,50 +39,51 @@ internal static string GetMessage(int errorCode)
...
             int flags = FORMAT_MESSAGE_IGNORE_INSERTS | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ARGUMENT_ARRAY;
             if (moduleHandle != IntPtr.Zero)
             {
                 flags |= FORMAT_MESSAGE_FROM_HMODULE;
             }
 
-            int result = FormatMessage(flags, moduleHandle, (uint)errorCode, 0, sb, sb.Capacity, null);
-            if (result != 0)
+            int length;
+            unsafe
```
Comment (danmosemsft) [inline src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: Seems a shame it has to use unsafe code.

Comment (stephentoub) [reply src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: To be clear, this code is all pre-existing this change; you can see it in the file I'm deleting (all I did was `move /Y corefx\src\Common\Interop\path corefx\src\Common\CoreLib\Interop\path`.

That said, what would you recommend instead?

File: src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs (hunk 1 above)
Comment (jkotas) [reply src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: Rare case on error path. Not worth it.

Comment (jkotas) [reply src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: If we wanted to do something about this, using `FORMAT_MESSAGE_ALLOCATE_BUFFER` and avoiding the retry logic completely would be best.

File: src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs [hunk 4]
```diff
@@ -105,8 +107,5 @@ private static bool TryGetErrorMessage(IntPtr moduleHandle, int errorCode, Strin
         //
```
Comment (jkotas) [inline src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: This comment should stay with the constants.

Comment (jkotas) [reply src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: > the buffer can only be freed using LocalFree and it is a private API on onecore.

This part of the comment is no longer valid. LocalFree API is available even in UAP.

Comment (stephentoub) [reply src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.FormatMessage.cs]: Ah... then there's no point in doing the whole growth allocation approach, either.  I can change it to start with the stack memory, and then if that's insufficient, ask FormatMessage to allocate and then I can free it.

**Comment by riarenas** [comment]
@stephentoub Opened dotnet/core-eng#3455 to look into the osx failures.

**Comment by stephentoub** [comment]
I updated it to use FORMAT_MESSAGE_ALLOCATE_BUFFER instead of looping with a growable buffer.

**Comment by stephentoub** [comment]
Thanks, Drew.

@dotnet-bot test this please

**Pull request merged**